pkg/discovery/util/node_util.go

```diff
@@ -217,3 +218,43 @@ func DetectHARole(node *api.Node) bool {
 	}
 	return isHANode
 }
+
+func MapNodePoolToNodes(nodes []*v1.Node, machineSetToNodesMap map[string][]*v1.Node) map[string][]*v1.Node {
+	glog.V(3).Info("mapping node pools to nodes.")
+	nodePools := make(map[string][]*v1.Node)
+	for _, node := range nodes {
+		allPools := DetectNodePools(node)
+		for _, pool := range allPools.List() {
+			nodePools[pool] = append(nodePools[pool], node)
+		}
+	}
+
+	for capiMachineSetPoolName, nodes := range machineSetToNodesMap {
+		nodePools[capiMachineSetPoolName] = nodes
+	}
+
+	glog.V(3).Infof("Found %+v node pool keys.", len(nodePools))
+	glog.V(4).Info(nodePools)
+	return nodePools
+}
+
+type NodeLike interface {
+	string | *v1.Node
+}
+
+func IsSuperset[T NodeLike](nodeNames sets.String, nodes []T, extractor func(T) string) bool {
+	for _, node := range nodes {
+		if !nodeNames.Has(extractor(node)) {
+			return false
+		}
+	}
+	return true
+}
```

Comment:
This is interesting, I have never used Generics with Go. Just out of curiosity, why do we need to have a generic type here if the only type we need is `*v1.Node`? 

By the way, thanks for checking super set, I think this is better than doing equal.

Reply:
So I was originally using this function for checking node names in a set with node names from an slice, but then I needed it for a node slice so I figured I would keep the generic version. Over time we might even use something like a Byte[] or NodeMap.